[MIOpen] Ported GRU and GRU dropout test from CTest to GTest

dpolominDXC opened the pull request:
## Motivation

Ongoing porting from CTest to GTest

## Technical Details

GRU and GRU dropout used the same class derived from gru_basic_driver and it made sense to combine them into one GTest cpp file, because of identical code. I also had some trouble with test tolerance, which seems to be pretty high, but the same was for the original test.
I also removed copies of CPU solvers and used the ones from rnn_driver. I see no difference in results between them and the code is almost identical.

## Test Plan

This is the test itself, so just launched it. Also, the CI job will be launched for this test, obviously

## Test Result

See the relevant CI job

## Submission Checklist

- [ ] Look over the contributing guidelines at https://github.com/ROCm/ROCm/blob/develop/CONTRIBUTING.md#pull-requests.

dpolominDXC committed:
Ported GRU and GRU dropout test from CTest to GTest

dpolominDXC committed:
Fixed GRU headers in tests

Vsevolod1983 commented:
@shurale-nkn 
Please make a review!

shurale-nkn commented on projects/miopen/test/gtest/gru_test.cpp:
```diff
@@ -0,0 +1,1618 @@
+        {
+// Workaround for issue #2335.
+// OpenCL error creating buffer: 0 Invalid Buffer Size
+#if MIOPEN_BACKEND_OPENCL
```
need to remove. we don't have BACKEND_OPENCL test, opencl not used in gtests

dpolominDXC replied:
Removed

shurale-nkn commented on projects/miopen/test/gtest/gru_test.cpp:
```diff
@@ -0,0 +1,1618 @@
+            unsigned long long dropout_seed = 0ULL;
+            miopenDropoutGetStatesSize(mio_handle, &statesSizeInBytes);
+
+#if MIOPEN_BACKEND_OPENCL
```
same need to remove

dpolominDXC replied:
Done

shurale-nkn commented on projects/miopen/test/gtest/deepbench_gru.cpp:
```diff
@@ -25,7 +25,7 @@
 #include <miopen/miopen.h>
 #include <gtest/gtest.h>
-#include "../gru.hpp"
+#include "gru.hpp"
```
?? Please clarify how this works, I think you deleted this file.

dpolominDXC replied:
Yes, I did, thanks for pointing it out. I didn't check those files and built only one test. Now I moved test cases from this file into main one and made it its own test

shurale-nkn commented on projects/miopen/test/gtest/gru_extra.cpp:
```diff
@@ -25,7 +25,7 @@
 #include <miopen/miopen.h>
 #include <gtest/gtest.h>
-#include "../gru.hpp"
+#include "gru.hpp"
```
same

dpolominDXC replied:
Done

shurale-nkn commented on projects/miopen/test/gtest/gru_test.cpp:
```diff
@@ -0,0 +1,1618 @@
+                        &(miopen::deref(miopen::deref(rnnDesc).dropoutDesc)),
+                        nohx);
+
+#if(MIO_GRU_TEST_DEBUG == 2)
```
I think we cand remove this old code

dpolominDXC replied:
Done

shurale-nkn commented on projects/miopen/test/gtest/gru_test.cpp:
```diff
@@ -0,0 +1,1618 @@
+template<typename T>
+struct GruTolerance
+{
+    static double value() { return 500000.f; }
```
looks huge ... original tolerance 80 * eps

dpolominDXC replied:
Yeah. Turned out, I had several problems with it. You may notice I'm using my own header for simplicity called `compare_helper.hpp`. Turns out, I made a 'mistake' there by executing `gpu()` first and `cpu()` second.
It shouldn't matter, but this code is written in such a way (maybe reusing buffers or just essentially relying on side effects), that when I noticed that `test_driver` class did it the other way around and swapped them, I got the tolerance issue resolved.
Not to mention that removing this test CPU solvers and trying to reuse already existing ones also contributed to big difference...
So, now that I have copied CPU solvers from old CTest and also have `gpu()` executing first and `cpu()` second, I finally have tolerance as expected. Took me just 3 weeks...

dpolominDXC committed:
Changed order in compare_helper to better emulate previous test_driver
implementation
Returned to GRU test's own CPU solver functions, since their side
effects are being relied on
Used hardcoded test cases as 'Full' tests instead of trying to replicate obscure
test_driver's run_data behaviour
Brought over gru_deepbench and gru_extra as test cases

dpolominDXC committed:
Updated license for compare_helper.hpp

dpolominDXC commented:
I had a lot of problems trying to create 64 test cases from 15 parameters like original CTest did via super-obscure `run_data` function and `argument` struct that contains lambdas embedded into lambdas.
So, instead of trying to replicate this piece of pure coding art (sorry, I had hard time reading it), I just hard-coded the test cases into this test that were produced by the CTest. Thankfully, there's not too many.

dpolominDXC committed:
Removed redundant code from gru_test

dpolominDXC committed:
Fixed GRU smoke test generation

dpolominDXC committed:
Applied format to GRU test

dpolominDXC committed:
Removed redundand assignment in GRU smoke test generation

Vsevolod1983 commented on projects/miopen/test/gtest/compare_helper.hpp:
```diff
@@ -112,8 +89,8 @@ template <class VerifyT>
 {
-    const auto cpu_result = verifier.cpu();
     const auto gpu_result = verifier.gpu();
+    const auto cpu_result = verifier.cpu();
```
Could you please clarify, why did you change the order here?
In case if it is important for GRU test, I think it's worth to make its own version of CompareResults function (and please add a comment on why you are doing that).
This function is common (is used by all tests), so if anybody change this order back it would fail GRU.

dpolominDXC replied:
Uhmm, I was the one who created the file (Djordje ported it from my unfinished PR) and it was intended to quickly mimic whatever `test_driver` did. There are mostly my tests which use it, and in their case I'm sure they will be fine.

I made a mistake there: `test_driver` actually launched `gpu()` first and if verification flag is on, then it launched `cpu()` function in another thread but immediately started waiting on it. That means it's essentially linear, despite a separate thread.
So, basically, here I restored the initial `test_driver` order.

Now, it shouldn't really matter in which order you launch those, as you pointed out. But in case of this test, it somehow does (like reusing the buffers or something), and I really don't feel I should investigate it. It is how it was written, so if you're investigating it, then you most likely will completely refactor it (which I did initially, and it produced even more strange errors).

dpolominAMD replied:
We've discussed separately, that it would be better not to touch this file and make a separate implementation of this `Compare` function for this particular test. This'll highlight its unusual behaviour

Vsevolod1983 commented on projects/miopen/test/gtest/gru_test.cpp:
```diff
@@ -2845,76 +2602,553 @@ struct verify_backward_weights_gru
+};
+
+std::vector<TestCase> dropout_full_cases = {
+    TestCase{17, 23, 13, 67, 3, 0, 0, 0, true, false, false, false, false, false, {0}},
```
please use Combine mechanism for generating test cases

dpolominDXC replied:
OK, fixed those, I hope. Let's see the build, though...
I hope you'll forgive me leaving 'bench' and 'extra' tests as hardcoded, because they were like that in the first place.

Vsevolod1983 commented on projects/miopen/test/gtest/gru_test.cpp:
```diff
@@ -2845,76 +2602,553 @@ struct verify_backward_weights_gru
+    TestCase{32, 3, 128, 128, 1, 0, 0, 1, false, true, true, true, true, false, {32, 32, 32}}};
+
+static std::vector<TestCase> base_cases = {
+    TestCase{1, 1, 13, 67, 1, 0, 0, 0, false, false, false, false, false, false, {0}},
```
I can see "add" method is used in the original ctest in order to produce all possible combination. Please use Combine mechanism (or custom loops if needed). It gives better understandings on how we generate test cases and will make it easier to add new ones

dpolominDXC replied:
Same as previous comment

Vsevolod1983 commented on projects/miopen/test/gtest/gru_test.cpp:
```diff
@@ -3226,4 +3470,135 @@ struct gru_basic_driver : test_driver
+using GPU_GRU_Extra_FP32 = gru_extra_test<float>;
+using GPU_GRU_Extra_FP16 = gru_extra_test<half_float::half>;
+
+TEST_P(GPU_GRU_Base_FP32, TestFloat32) { Run(); }
```
Could you please check which types are possible for gru basic? Is it only fp32 and fp16 or there is something else ?

dpolominDXC replied:
Yes, I did check it and it says 'RNN datatype must be float or half':
```
MIOpen Error: 5d738dde443a:/data/rocm-libraries-second/projects/miopen/src/rnn.cpp:333: RNNDescriptor: Bad parameter(s). RNN datatype must be float or half.
test_gru: /data/rocm-libraries-second/projects/miopen/src/include/miopen/handle.hpp:206: Allocator::ManageDataPtr miopen::Handle::Write(const Container &) const [Container = std::vector<double>]: Assertion `!c.empty()' failed.
Aborted (core dumped)
```

dpolominDXC committed:
Used generation functions instead of hardcoded cases for gru test

dpolominDXC committed:
GRU test: fixing warning

dpolominDXC committed:
Reverted changes in compare_helper.hpp and left GRU tests with its own implementation
